### ISSUE TITLE
Option to provide a list of non-shrunken parameters

### ISSUE THREAD

**Issue by michael-brent** [BrentLab/tfbpmodeling#19]
Zolboo is working on this. Brief motivation: We want to explain as much of the variance in the data as possible using a non-linear fit to the perturbed TF, so that any interaction terms we select are not just there as proxies for non-linear effects of the pTF. Therefore, we may not want to apply shrinkage to the binding of the pTF, its square, and its cube. This basically gives these parameters free reign to do as much as they can, and the interaction terms will have to pick up whatever is left over. I expect this will reduce the number of interaction terms kept, but in combination with the new new stepwise CI elimination, we may still have a lot.

Expose to the user the option t provide a list of coefficients that should not be included in the shrinkage. As a result, I expect that these coefficients will always pass the bootstrap CI.

**Comment by cmatKhan** [comment]
@michael-brent , I looked through the LassoCV source code from sklearn. The function used to call the coordinate descent is `path`. The Lasso path is called `lasso_path`, which calls `enet_path` (elastic net path) with the L2 penalty set to 0.

That brings us to a Cython function, which is a performant version of python, which conducts the gradient descent. This is where the penalty is calculated:

https://github.com/scikit-learn/scikit-learn/blob/90209c8cdda28462b0c199f3704721b5cc9923c9/sklearn/linear_model/_cd_fast.pyx#L877

```python
 _scal(n_tasks, fmax(1. - l1_reg / nn, 0) / (norm_cols_X[ii] + l2_reg), &W[0, ii], 1)
```

I should thank chatGPT at this point -- I wouldn't have known how to figure out what `_scal` is. It is a call to a [Blas subroutine](https://icl.bitbucket.io/blaspp/group__scal.html).

<h3>_scal</h3>

<p>Scale vector by constant, \(x = \alpha x\). </p>
<p>Generic implementation for arbitrary data types.</p>
<dl class="params"><dt>Parameters</dt><dd>
  
[in] | n | Number of elements in x. n >= 0.
-- | -- | --
[in] | alpha | Scalar alpha.
[in] | x | The n-element vector x, in an array of length (n-1)*incx + 1.
[in] | incx | Stride between elements of x. incx > 0.

</dd></dl>

`fmax` is a call to a C function which takes the maximum between two values, and `l1_reg` is, for us since we're doing Lasso, 1. This is the elastic net coord descent routine, but it can of course be used for Lasso or the L2 penalty by setting the `l1_reg` to 1 or 0, for Lasso or Ridge regression respectively.

To remove certain features from the L1 penalty, we multiply the penalty term by a coef_penalty_mask[ii], which is 0 for unpenalized features and 1 otherwise (again, thank you chatGPT):

```python
_scal(n_tasks, fmax(1. - (l1_reg * coef_penalty_mask[ii]) / nn, 0) / (norm_cols_X[ii] + l2_reg),  &W[0, ii], 1)
```


For the third time, ChatGPT helped significantly with this -- here is a suggested (and untested) implementation outline:

```python
def coef_weighted_enet_coordinate_descent_multi_task(
    const floating[::1, :] W,
    floating l1_reg,
    floating l2_reg,
    const floating[::1, :] X,
    const floating[::1, :] Y,
    unsigned int max_iter,
    floating tol,
    object rng,
    bint random=0,
    const floating[::1,:] coef_penalty_mask,
):
# note that coef_penalty_mask is added as an argument
...
```

Then we'd need to modify the `enet_path` so that `coef_penalty_mask` could be passed through `params` (no modification necessary to the signature or the bulk of the function

```python
def enet_path(
    X,
    y,
    *,
    l1_ratio=0.5,
    eps=1e-3,
    n_alphas=100,
    alphas=None,
    precompute="auto",
    Xy=None,
    copy_X=True,
    coef_init=None,
    verbose=False,
    return_n_iter=False,
    positive=False,
    check_input=True,
    **params,
):
...
    coefs, gaps = coef_weighted_enet_coordinate_descent_multi_task(
        X=X, y=y, l1_ratio=l1_ratio, alphas=alphas,
        penalty_mask=penalty_mask, **kwargs
    )
    return alphas, coefs, gaps
```

Then we can write the custom LassoCV class

```python
from sklearn.linear_model import LassoCV
from sklearn.linear_model._coordinate_descent import _path_residuals

class CustomLassoCV(LassoCV):
    def __init__(self, *, penalty_mask=None, **kwargs):
        super().__init__(**kwargs)
        self.penalty_mask = penalty_mask

    def _path_residuals(self, X, y, train, test, alphas, l1_ratio, X_order, dtype):
        return _path_residuals(
            X=X, y=y, train=train, test=test,
            fit_intercept=self.fit_intercept,
            path=my_enet_path,
            path_params={
                "l1_ratio": l1_ratio,
                "penalty_mask": self.penalty_mask,
                "fit_intercept": self.fit_intercept,
                "normalize": False,
                "copy_X": True,
                "verbose": self.verbose,
                "max_iter": self.max_iter,
                "tol": self.tol,
            },
            alphas=alphas,
            l1_ratio=l1_ratio,
            X_order=X_order,
            dtype=dtype,
        )
```

And a minimal test

```python
def test_masked_features_not_penalized():
    from sklearn.datasets import make_regression
    X, y = make_regression(n_samples=100, n_features=5, noise=0.1)
    mask = [0, 0, 1, 1, 1]  # First two features unpenalized

    model = CustomLassoCV(penalty_mask=mask, alphas=[0.1])
    model.fit(X, y)

    # Check that the unpenalized weights are relatively large
    assert abs(model.coef_[0]) > 0.1
    assert abs(model.coef_[1]) > 0.1
```

**Comment by cmatKhan** [comment]
This may be relevant

**Comment by cmatKhan** [comment]
Even better, this is weighted lasso CV implemented by Microsoft Research

https://econml.azurewebsites.net/_autosummary/econml.sklearn_extensions.linear_model.WeightedLassoCV.html

https://github.com/py-why/EconML/blob/main/econml/sklearn_extensions/linear_model.py#L331

**Comment by cmatKhan** [comment]
https://github.com/alvaromc317/asgl

that is probably worth looking at -- it does implement adaptive lasso in python

**Comment by cmatKhan** [comment]
this is available here

https://github.com/cmatkhan/glmnetstrap
